packages/protocol/contracts/stability/Reserve.sol

```diff
@@ -302,10 +307,50 @@ contract Reserve is
    * @param spender The address that is to be no longer allowed to spend Reserve funds.
    */
   function removeSpender(address spender) external onlyOwner {
-    isSpender[spender] = false;
+    delete isSpender[spender];
     emit SpenderRemoved(spender);
   }
 
+  function isAllowedToSpendExchange(address spender) public view returns (bool) {
+    return
+      isExchangeSpender[spender] || (registry.getAddressForOrDie(EXCHANGE_REGISTRY_ID) == spender);
+  }
+
+  /**
+   * @notice Gives an address permission to spend Reserve without limit.
+   * @param spender The address that is allowed to spend Reserve funds.
+   */
+  function addExchangeSpender(address spender) external onlyOwner {
+    require(!isExchangeSpender[spender], "Address is already Exchange Spender");
+    isExchangeSpender[spender] = true;
+    exchangeSpenderAddresses.push(spender);
+    emit ExchangeSpenderAdded(spender);
+  }
+
+  /**
+   * @notice Takes away an address's permission to spend Reserve funds without limits.
+   * @param spender The address that is to be no longer allowed to spend Reserve funds.
+   */
+  function removeExchangeSpender(address spender, uint256 index) external onlyOwner {
+    isExchangeSpender[spender] = false;
+    uint256 numAddresses = exchangeSpenderAddresses.length;
+    require(index < numAddresses, "Index is invalid");
+    require(spender == exchangeSpenderAddresses[index], "Index does not match spender");
+    uint256 newNumAddresses = numAddresses.sub(1);
+
+    if (index != newNumAddresses) {
+      exchangeSpenderAddresses[index] = exchangeSpenderAddresses[newNumAddresses];
+    }
+
+    exchangeSpenderAddresses[newNumAddresses] = address(0x0);
+    exchangeSpenderAddresses.length = newNumAddresses;
+    emit ExchangeSpenderRemoved(spender);
+  }
+
+  function getExchangeSpenders() public view returns (address[] memory) {
```

Comment:
Can we remove this? `exchangeSpenderAddresses` is a public variable so this is redundant

Reply:
Looks like this is only used for testing-- think it's worth removing? iirc there isn't a clean way of getting `exchangeSpenderAddresses.length` from web3 to properly iterate through. Because it's already in there for CR3 I don't think we would gain much now from removing it, if anything it would introduce backward incompatibility